src/app/tests/TestReadInteraction.cpp

```diff
@@ -0,0 +1,190 @@
+};
+
+void TestReadInteraction::GenerateReportData(nlTestSuite * apSuite, void * apContext, System::PacketBufferHandle aPayload)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    System::PacketBufferTLVWriter writer;
+    writer.Init(std::move(aPayload));
+
+    ReportData::Builder reportDataBuilder;
+
+    err = reportDataBuilder.Init(&writer);
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+    reportDataBuilder.SuppressResponse(true);
+    NL_TEST_ASSERT(apSuite, reportDataBuilder.GetError() == CHIP_NO_ERROR);
+
+    reportDataBuilder.MoreChunkedMessages(false);
+    NL_TEST_ASSERT(apSuite, reportDataBuilder.GetError() == CHIP_NO_ERROR);
+
+    reportDataBuilder.EndOfReportData();
+    NL_TEST_ASSERT(apSuite, reportDataBuilder.GetError() == CHIP_NO_ERROR);
+
+    err = writer.Finalize(&aPayload);
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+}
+
+void TestReadInteraction::TestReadClient(nlTestSuite * apSuite, void * apContext)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    app::ReadClient readClient;
+
+    System::PacketBufferHandle buf = System::PacketBufferHandle::New(System::PacketBuffer::kMaxSize);
+    err                            = readClient.Init(&gExchangeManager, nullptr);
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+    err = readClient.SendReadRequest(kTestDeviceNodeId, gAdminId);
+    NL_TEST_ASSERT(apSuite, err == CHIP_ERROR_INCORRECT_STATE);
+
+    GenerateReportData(apSuite, apContext, buf.Retain());
+
+    err = readClient.ProcessReportData(std::move(buf));
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+    readClient.Shutdown();
+}
+
+void TestReadInteraction::TestReadHandler(nlTestSuite * apSuite, void * apContext)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    app::ReadHandler readHandler;
+    System::PacketBufferTLVWriter writer;
+    System::PacketBufferHandle reportDatabuf  = System::PacketBufferHandle::New(System::PacketBuffer::kMaxSize);
+    System::PacketBufferHandle readRequestbuf = System::PacketBufferHandle::New(System::PacketBuffer::kMaxSize);
+    ReadRequest::Builder readRequestBuilder;
+    readHandler.Init(&gExchangeManager, nullptr);
+
+    GenerateReportData(apSuite, apContext, reportDatabuf.Retain());
+    err = readHandler.SendReportData(std::move(reportDatabuf));
+    NL_TEST_ASSERT(apSuite, err == CHIP_ERROR_INCORRECT_STATE);
+
+    writer.Init(std::move(readRequestbuf));
+    err = readRequestBuilder.Init(&writer);
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+    readRequestBuilder.EventNumber(1);
+    NL_TEST_ASSERT(apSuite, readRequestBuilder.GetError() == CHIP_NO_ERROR);
+    readRequestBuilder.EndOfReadRequest();
+    NL_TEST_ASSERT(apSuite, readRequestBuilder.GetError() == CHIP_NO_ERROR);
+    err = writer.Finalize(&readRequestbuf);
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+    err = readHandler.OnReadRequest(nullptr, std::move(readRequestbuf));
```

Comment:
This is going to stop working once the read handler is actually sending responses, no?  Followup ok to sort that out, but maybe add a TODO comment here to make it clear that this is known problem?

Reply:
in #4800, once OnReadRequest process read request, it would schedule report engine to send out report data, so this code would still work. 
https://github.com/project-chip/connectedhomeip/pull/4800/files#r595739935

Reply:
But isn't there supposed to be an immediate response on the _same exchange_?  Where is that "same exchange" going to come from?

Reply:
The design is as below.
1. When Interaction Model Engine receives read request, it will call onReadRequest and save the current exchange context into ReadHandler. When it call report engine' run, report engine run would check all active read handler with exchange context, and instruct these active read handlers(possibly multiple read requests in flight) to fetch interested attribute/event path lists and send out them to the initiator.  

2. When considering subscribe request, it can reuse the above model and code, when there exists multiple active subscription, whenever there is changes happening in device, it can trigger report engine run, and ask subscription handler to generate reports according to the interested path.

3. Stack Depth considering, Separation between onReadRequest or later onSubscriptionRequest and ReportGeneration from attribute/event paths via Invoking ScheduleRun could reduce the stack depth in limited embedded device.